### PR TITLE
Python 3.4/5/6 and PyPy support

### PR DESCRIPTION
**Pull request by jamadden**
WIP because Python 3 tests won't reliably pass until https://github.com/buildout/buildout/pull/393 is released.

Also WIP because coverage isn't right yet because of all the subprocesses. I still need to work on that.

### CODE REVIEW

**Comment by jamadden** [comment]
I got coverage for the subprocesses figured out (a tiny bit of boiler code that could potentially be added to the existing buildoutSetUp/tearDown methods in zc.buildout.testng). It's pretty good at 96%. 

It's still a WIP pending a zc.buildout release.

File: setup.py [hunk 1]
```diff
@@ -53,16 +64,29 @@ def read(*rnames):
...
-    data_files = [('.', ['README.txt'])],
-    namespace_packages = ['zc', 'zc.recipe'],
-    install_requires = ['zc.buildout >=1.4', 'setuptools'],
-    extras_require = dict(test=['zope.testing']),
-    entry_points = {'zc.buildout':
-                    ['default = %s:Recipe' % name]},
-    zip_safe = True,
-    )
+    ),
+    package_dir={'':'src'},
```
Comment (mgedmin) [inline setup.py]: Space after the `:` please?

File: setup.py [hunk 2]
```diff
@@ -53,16 +64,29 @@ def read(*rnames):
...
+    namespace_packages=['zc', 'zc.recipe'],
+    install_requires=[
+        'zc.buildout >= 1.4',
+        'setuptools'],
+    extras_require={
+        'test': [
+            # sadly zc.buildout doesn't have a test extra, so we
+            # need to duplicate its test dependencies, since we import its
+            # test package.
+            'zc.buildout[test]',
```
Comment (mgedmin) [inline setup.py]: You say that zc.buildout doesn't have a test extra, and then go ahead and use zc.buildout's test extra.  ... ?

File: setup.py [hunk 3]
```diff
@@ -16,35 +16,46 @@
 from setuptools import setup, find_packages
 
 def read(*rnames):
-    return open(os.path.join(os.path.dirname(__file__), *rnames)).read()
+    with open(os.path.join(os.path.dirname(__file__), *rnames)) as f:
+        return f.read()
 
-name = "zc.recipe.cmmi"
+name="zc.recipe.cmmi"
```
Comment (mgedmin) [inline setup.py]: Please no?

File: src/zc/recipe/cmmi/tests.py [hunk 4]
```diff
@@ -21,49 +25,87 @@
...
+        # coverage config absolute first.
+        assert os.path.isfile(path_to_coveragerc), path_to_coveragerc
+        os.environ['COVERAGE_PROCESS_START'] = path_to_coveragerc
+        test.globs['_path_to_coveragerc'] = path_to_coveragerc
+
+        with open('bin/buildout') as f:
+            lines = f.read().splitlines()
+            assert lines[1] == '', lines
+            lines[1] = 'import coverage; coverage.process_startup()'
+        with open('bin/buildout', 'w') as f:
```
Comment (mgedmin) [inline src/zc/recipe/cmmi/tests.py]: O.O

...

I suppose it works :/

**Review by mgedmin — APPROVED** [review]
lgtm, I suppose.

I don't quite understand how you can keep re-writing bin/buildout without failing when the tests run again and line[1] no longer passes the == '' assertion, but let's assume I missed something and it's fine.

**Comment by mgedmin** [comment]
> Package Index Owner: J1m, batlogg, fdrake, tarek, jrush, tseaver, wosc, tlotze, patricks, davisagli
> Package Index Maintainer: batlogg, fdrake, davisagli

File: setup.py (hunk 2 above)
Comment (jamadden) [reply setup.py]: I'm wrong about that. It *does* have a test extra, but I can't figure out how to get the test extra to install when I install from the VCS url. 

But anyway, looking at the test extra it drags in things we really don't want ('bobo ==2.3.0', 'zdaemon', 'zc.zdaemonrecipe', 'zc.recipe.deployment') and that apparently aren't necessary.

**Comment by jamadden** [comment]
> I don't quite understand how you can keep re-writing bin/buildout without failing when the tests run again and line[1] no longer passes the == '' assertion

The `setUp` method defined here is used for doctests, and so is called only once for each file. The first thing it does is call `zc.buildout.testing.buildoutSetUp(test)` which creates a new temporary directory containing `bin/buildout` (and changes to it). Our `setUp` then proceeds to modify that file. This happens just once for the whole test file.

**Comment by jamadden** [comment]
Since we're going to require a new zc.buildout release *anyway*, I'd like to prepare a PR for it that moves the coverage support code there. Hopefully it will be accepted and make life easier for all buildout recipe authors that use this doctest-based testing that spawn subprocesses.

**Comment by jamadden** [comment]
buildout 2.9.4 has been released with the fix we needed and the support for coverage added. Tests are green here and PR has been approved, so I'm going to go ahead and merge.

@jimfulton Could we get a PyPI release please? Or if you'd like to add `jamadden` and `mgedmin` as PyPI owners I'd be happy to take care of it.

**Comment by tseaver** [comment]
I've added both 'jamadden' and 'mgedmin' as project owners on PyPI.

**Comment by jamadden** [comment]
Thank you!

**Pull request merged**